### DISCUSSION TITLE
Valid tags by level?

### DISCUSSION THREAD

**Discussion by Snake69** [FamilySearch/GEDCOM#203] [General]
Is there a definitive list of which tags are valid for level 0?

http://taggedwiki.zubiaga.org/new_content/27852e435f72b674140a32c714719e99
seems to indicate that the only valid level 0 tags are:
HEAD, TRLR, SUBN, INDI, FAM, OBJE, NOTE, REPO, SOUR, and SUBM

Is this correct?

Is there a definitive list of which level 1 tags are valid for each of the
different level 0 tags?

At https://www.genealogieonline.nl/en/GEDCOM-tags/?prog=all a file can be
downloaded containing stats of how often tags (and tag levels) were referenced
in a sample of 9700 GEDCOM files.  Looking through the file shows multiple
times where apparent invalid usages occurred (e.g., the tags BIRT and DEAT
were used as level 0 many times).

**Comment by tychonievich** [comment]
> Is there a definitive list of which tags are valid for level 0?

Yes. They are the Dataset and RECORD productions in the spec: <https://gedcom.io/specifications/FamilySearchGEDCOMv7.html#structure-organization>. The list you mention is correct for 5.5.1; in 7.0 NOTE is no longer level-0 but the new SNOTE is; and SUBN is no longer present.

> Is there a definitive list of which level 1 tags are valid for each of the different level 0 tags?

This is given in the specification, but not in list format because some use references to groups of level-1 tags shared by multiple structures.

A direct list of substructures of any structure can be found in the `substructures` field of the YAML file served at its URI; for example, <https://gedcom.io/terms/v7/record-SOUR>.

Both questions are also answered by the list of all (superstructure, tag, substructure) triples defined in the specification which is automatically updated with each release in the substructures TSV <https://github.com/FamilySearch/GEDCOM/blob/main/extracted-files/substructures.tsv>. That does not list the `HEAD` and `TRLR`, which are special metadata required of every file, but does list all of the other level-0 tags with a blank superstructure field.

You are correct that invalid files were common in implementations of 5.5 and 5.5.1. I believe this is partly because those standards were not accompanied by machine-parseable metadata like the YAML and TSV files noted above, making compliance much harder to verify.

As an aside, extension structures, whose tags start with an underscore (`_`) can in principle appear at any level and have any substructure. We're in the process of providing registries for their YAML files too so that they can also be verified as matching their individual extension specifications.

**Reply by bunahu** [reply]
I was expecting, intuitively, that GEDCOM would be built around individuals and/or families, but that's not correct, is it?

From what I can gather in the documentation and from your comment above, it's really built fundamentally around records (documents, multimedia, etc.). Every top-level tag is some sourced record _referencing_ a family/individual, but there's no top-level structure _defining_ a family/individual. So my entire "family tree" is built in GEDCOM-compatible software using:

<< HEADER >>
<< Record 1 >>
<< Record 2 >>
.
.
.
<< Record N >>
<< TRLR >>

And it's not a collection of people, but a collection of records.

Is this a correct (if simplified) view of what's going on in GEDCOM files - that "it's << RECORD >>s all the way down?" If so, I didn't get that from reading the documentation the past couple weeks, for whatever reason.

**Comment by tychonievich** [comment]
> I was expecting that GEDCOM would be built around individuals and/or families [...] it's really built fundamentally around records (documents, multimedia, etc.)

Your intuition that families and individuals are central is correct.

In what now appears like an unfortunate naming decision, from the beginning of GEDCOM's history level-0 structures were called "records". This was in keeping with the programming vernacular of the day: "record" was what Pascal and ALGOL called what C (standardized 2 years after GEDCOM 3.0) calls "structs" and what the later OO-languages call "objects" and "classes".

Most "records" under this naming have nothing to do with the genealogist's use of the word "record" to mean a historical document. The only place the two meanings of "record" line up is in the `SOUR` record, which often describes historical records, and in some cases the `OBJE` record, which may link to digital copies of historical records.

> it's << RECORD >>s all the way down?

No; the word "record" in GEDOCM is reserved for level-0 structures; larger levels are called "structures" in 7.0, and sometimes "structures" and sometimes "lines" in earlier versions. If you mean historical records, also no: it is a common lament that many creators of genealogical data provide far too few `SOUR`ces for the claims they make.

**Reply by Snake69** [reply]
Thanks, once again, for your help, Luther.  And, by the way, thanks for providing the nice GEDCOM7 JS parser.
